numpy/_core/src/multiarray/stringdtype/dtype.c

```diff
@@ -667,6 +680,165 @@ static PyType_Slot PyArray_StringDType_Slots[] = {
         {0, NULL}};
 
 
+/*
+ * Wrap the sort loop to acquire/release the string allocator.
+ */
+int
+stringdtype_wrap_sort_loop(
+        PyArrayMethod_Context *context,
+        char *const *data, const npy_intp *dimensions, const npy_intp *strides,
+        NpyAuxData *transferdata, PyArray_SortImpl *sort_loop)
+{
+    PyArray_StringDTypeObject *sdescr =
+            (PyArray_StringDTypeObject *)context->descriptors[0];
+
+    npy_string_allocator *allocator = NpyString_acquire_allocator(sdescr);
+    int ret = sort_loop(
+            data[0], dimensions[0], context,
+            context->descriptors[0]->elsize, &_sort_compare);
+    NpyString_release_allocator(allocator);
+    return ret;
+}
+
+/* 
+ * This is currently required even though the default implementation would work,
+ * because the output, though enforced to be equal to the input, is parametric.
+ */
+static NPY_CASTING
+stringdtype_sort_resolve_descriptors(
+        PyArrayMethodObject *method,
+        PyArray_DTypeMeta *const *dtypes,
+        PyArray_Descr *const *input_descrs,
+        PyArray_Descr **output_descrs,
+        npy_intp *view_offset)
+{
+    output_descrs[0] = NPY_DT_CALL_ensure_canonical(input_descrs[0]);
+    if (NPY_UNLIKELY(output_descrs[0] == NULL)) {
+        return -1;
+    }
+    output_descrs[1] = NPY_DT_CALL_ensure_canonical(input_descrs[1]);
+    if (NPY_UNLIKELY(output_descrs[1] == NULL)) {
+        Py_XDECREF(output_descrs[0]);
+        return -1;
+    }
+
+    // sorting cannot be a view
+    *view_offset = 0;
+
+    return method->casting;
+}
+
+int
+stringdtype_wrap_argsort_loop(
+        PyArrayMethod_Context *context,
+        char *const *data, const npy_intp *dimensions, const npy_intp *strides,
+        NpyAuxData *transferdata, PyArray_ArgSortImpl *argsort_loop)
+{
+    PyArray_StringDTypeObject *sdescr =
+            (PyArray_StringDTypeObject *)context->descriptors[0];
+
+    npy_string_allocator *allocator = NpyString_acquire_allocator(sdescr);
+    int ret = argsort_loop(
+            data[0], (npy_intp *)data[1], dimensions[0], context,
+            context->descriptors[0]->elsize, &_sort_compare);
+    NpyString_release_allocator(allocator);
+    return ret;
+}
+
+int
```

Comment:
I think we need only one loop (say, `stringdtype_sort_loop` and `stringdtype_argsort_loop` if we use `transferdata` -- see comment on `get_loop`.